frameworks.qmd

```diff
@@ -1,119 +1,1785 @@
+evaluate performance and gain insights into the models\' workings. These
+tools enable disciplined experimentation and debugging.
+
+##### Evaluation Metrics
 
-- Definition of ML Frameworks
-- What is an ML framework?
-- Why are ML frameworks important?
-- Go over the design and implementation
-- Examples of ML frameworks
-- Challenges of embedded systems
+Frameworks include implementations of common evaluation metrics for
+validation:
 
-## Evolution of AI Frameworks
+-   Accuracy - Fraction of correct predictions overall. Widely used for
+    > classification.
 
-- High-level vs. low-level frameworks
-- Static vs. dynamic computation graph frameworks
-- Plot showing number of different frameworks and shrinking
+-   Precision - Of positive predictions, how many were actually
+    > positive. Useful for imbalanced datasets.
 
-## Types of AI Frameworks
+-   Recall - Of actual positives, how many did we predict correctly.
+    > Measures completeness.
 
-- Cloud-based AI frameworks
-- Edge AI frameworks
-- TinyML frameworks
+-   F1-score - Harmonic mean of precision and recall. Combines both
+    > metrics.
 
-## Popular AI Frameworks
+-   AUC-ROC - Area under ROC curve. Used for classification threshold
+    > analysis.
 
-Explanation: Discuss the most common types of ML frameworks available and provide a high-level overview, so that we can set into motion what makes embedded ML frameworks unique.
+-   MAP - Mean Average Precision. Evaluates ranked predictions in
+    > retrieval/detection.
 
-- TensorFlow, PyTorch, Keras, ONNX Runtime, Scikit-learn
-- Key Features and Advantages
-- API and Programming Paradigms
-- Table comparing the different frameworks
+-   Confusion Matrix - Matrix that shows the true positives, true
+    > negatives, false positives, and false negatives. Provides a more
+    > detailed view of classification performance.
 
-## Basic Components
+These metrics quantify model performance on validation data for
+comparison.
 
-- Computational graphs
-- Tensor data structures
-- Distributed training
-- Model optimizations
-- Code generation
-- Differentiable programming
-- Hardware acceleration support (GPUs, TPUs)
+##### Visualization
+
+Visualization tools provide insight into models:
+
+-   Loss curves - Plot training and validation loss over time to spot
+    > overfitting.
+
+![](images_ml_frameworks/image7.png){width="3.5052088801399823in"
+height="2.3043503937007874in"}
+
+Example of a loss curve from TensorBoard
+
+Source:
+[[https://www.tensorflow.org/tensorboard/scalars_and_keras]{.underline}](https://www.tensorflow.org/tensorboard/scalars_and_keras)
+
+-   Activation grids - Illustrate features learned by convolutional
+    > filters.
+
+-   Projection - Reduce dimensionality for intuitive visualization.
+
+-   Precision-recall curves - Assess classification tradeoffs.
+
+Tools like TensorBoard (TensorFlow) and TensorWatch (PyTorch) enable
+real-time metrics and visualization during training.
```

Comment:
Weights and Biases may also be a good reference for this! [W&B link](https://wandb.ai/site)

Reply:
That's a great suggestion, we should definitely add that.